File: lib/java/src/main/java/com/dianping/cat/message/internal/DefaultEvent.java
package com.dianping.cat.message.internal;

import com.dianping.cat.message.Event;
import com.dianping.cat.message.spi.MessageManager;

public class DefaultEvent extends AbstractMessage implements Event {
    private MessageManager manager;

    public DefaultEvent(String type, String name) {
        super(type, name);
    }

    DefaultEvent(String type, String name, MessageManager manager) {
        super(type, name);

        this.manager = manager;
    }

    @Override
    public void complete() {
        setCompleted(true);

        if (manager != null) {
            manager.add(this);
        }
    }
}

File: lib/csharp/src/CatClient/Message/Internals/DefaultEvent.cs
using System;
using Org.Unidal.Cat.Message.Spi;

namespace Org.Unidal.Cat.Message.Internals
{
    public class DefaultEvent : AbstractMessage, IEvent
    {
        private IMessageManager _mManager;

        public DefaultEvent(String type, String name) : base(type, name)
        {
        }

        public DefaultEvent(string type, string name, IMessageManager manager) : base(type, name)
        {
            _mManager = manager;
        }

        public override void Complete()
        {
            try
            {
                base.Complete();

                if (_mManager != null && _mManager.ThreadLocalMessageTree.Message == null)
                {
                    _mManager.Add(this);
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }
    }
}